src/NuGet.Core/NuGet.DependencyResolver.Core/ResolverUtility.cs

```diff
@@ -211,17 +211,21 @@ public static GraphItem<RemoteResolveResult> CreateUnresolvedMatch(LibraryRange
                     // check for the exact library through local repositories
                     var localMatch = await FindLibraryByVersionAsync(library, framework, localProviders, cacheContext, logger, cancellationToken);
 
-                    if (localMatch != null)
+                    if (localMatch != null
+                        && localMatch.Library.Version.Equals(library.Version))
```

Comment:
Version read lock file is exclusive, meaning this is the Version, not VersionRange which is why we create libraryIdentity and libraryRange in this case. But the resolution is for libraryRange which is why we need to make sure the resolved version is same as locked one. 

Other option is to come up with new Resources and new APIs which works for LibraryIdentity and only resolve to a specific version or else fail. This seems unnecessary at this time and only adds overhead. There is another PR# https://github.com/NuGet/NuGet.Client/pull/2578 which will help improve perf for these cases.

Reply:
Yeah I get that.

But if the version provided from the lock file is not [1.0.0], then wouldn't we have a problem with regular restore?

Reply:
There is an implicit operator that creates a LibraryRange (which is used in the selection APIs) from the LibraryIdentity you get from the lock file. 

https://github.com/NuGet/NuGet.Client/blob/7a5c6ae18ae07c22a6e3ad46f18f558a75830955/src/NuGet.Core/NuGet.LibraryModel/LibraryIdentity.cs#L76-L87

If you change the conversion from library identity to library range before you do the selection you can avoid the after the fact check. 

By using the implicit operator, there already is a conversion so you wouldn't be creating extra garbage. 

```
new VersionRange(
                            minVersion: library.Version,
                  includeMinVersion: true,
                  maxVersion: library.Version,
                  includeMaxVersion: true,
                  originalString: null,
                  floatRange: floatRange
                            )
            };
```

Reply:
Happy to discuss if my comment is confusing. 

There's a small performance advantage to my proposal because you'd save a few extra calls if the specific package versions appears only a specific source. That is probably a common case though. (see how we build against prerelease versions of VS/.NET Core dependencies which not found on nuget.org despite the other packages themselves being there).

Reply:
Also the implicit operator for LibraryIdentity -> LibraryRange is pretty bad in my opinion as it makes an incorrect assumption. The identity model cannot store all the information.

Reply:
I didn't want to change the implicit conversion because it's also being used in other places where it's fine to treat it as versionRange. 

There is no extra garbage, once it resolves a versionRange, it caches the results so next time it just returns the cached results. 

Sorry, but I also don't see any performance advantage what so ever, may be I'm missing something here... If you're talking about [this](https://github.com/NuGet/NuGet.Client/pull/2583/files#diff-ab60ca67cbc52f0cd86ebbd3df3441e0L440) then it's still applicable for libraryRange.

Again I failed to find the problem with implicit operator from LibraryIdentity to LibraryRange… It'd have been bad then it was other way around.

Last, we can further discuss it in person, or you can also share me a scenario where the new proposal is having better performance...

Reply:
Yeah, that's what I was referring to.

> There is no extra garbage, once it resolves a versionRange, it caches the results so next time it just returns the cached results.

I'm not saying it does, I was just saying it's the same amount of garbage :)

> Again I failed to find the problem with implicit operator from LibraryIdentity to LibraryRange… It'd have been bad if it was other way around.

The identity does not contain information about the max version of it. The implicit operator makes an assumption that versions always been float up if the specified version is not found. That's why I'm saying is bad. I'm not asking you to change it however, but merely to use a different conversion for the lock file case. 

It's alright thoguh, don't want to dwell on this for too long. 
I just think it's always make sense to check as many conditions as possible early rather than fragment the code like that. I think it reads better.

Feel free to merge. this one.